src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1817,107 +1817,123 @@ Node TheoryStringsRewriter::rewriteContains( Node node ) {
+  Node len0m2 = nm->mkNode(kind::MINUS, len0, node[2]);
+  if (checkEntailArith(len1, len0m2, true))
+  {
+    Node negone = nm->mkConst(Rational(-1));
+    return returnRewrite(node, negone, "idof-len");
+  }
+
+  Node fstr = node[0];
+  if (!node[2].isConst() || node[2].getConst<Rational>().sgn() != 0)
+  {
+    fstr = nm->mkNode(kind::STRING_SUBSTR, node[0], node[2], len0);
+    fstr = Rewriter::rewrite(fstr);
+    if (fstr.isConst())
+    {
+      CVC4::String fs = fstr.getConst<String>();
+      if (fs.size() == 0)
+      {
+        Node negone = nm->mkConst(Rational(-1));
+        return returnRewrite(node, negone, "idof-base-len");
       }
-      //decrement the start index
-      if( start>0 ){
-        if( s.size()>start ){
-          start = 0;
-        }else{
-          start = start - s.size();
+    }
+  }
+
+  Node cmp_con = nm->mkNode(kind::STRING_STRCTN, fstr, node[1]);
+  Node cmp_conr = Rewriter::rewrite(cmp_con);
+  if (cmp_conr.isConst())
+  {
+    if (cmp_conr.getConst<bool>())
+    {
+      if (node[2].isConst() && node[2].getConst<Rational>().sgn() == 0)
+      {
+        // pass first definite contain, we can drop
+        std::vector<Node> children1;
+        getConcat(node[1], children1);
+        std::vector<Node> nb;
+        std::vector<Node> ne;
+        int cc = componentContains(children0, children1, nb, ne, true, 1);
+        if (cc != -1 && !ne.empty())
+        {
+          Node nn = mkConcat(kind::STRING_CONCAT, children0);
+          Node ret = nm->mkNode(kind::STRING_STRIDOF, nn, node[1], node[2]);
+          return returnRewrite(node, ret, "idof-def-ctn");
         }
       }
-    }else if( !node[2].isConst() ){
-      break;
-    }else{
-      if( children[i]==node[1] && start==0 ){
-        //can remove beyond this
-        do_splice = true;
-      }
-    }
-    if( do_splice ){
-      std::vector< Node > spl;
-      //since we definitely will find the string, we can safely add the length of the constant non-overlapping prefix
-      if( prefixNoOverlap ){
-        Node pl = Rewriter::rewrite( NodeManager::currentNM()->mkNode( kind::STRING_LENGTH, children[0] ) );
-        Assert( pl.isConst() );
-        Assert( node[2].isConst() );
-        int new_start = val2 - pl.getConst<Rational>().getNumerator().toUnsignedInt();
-        if( new_start<0 ){
-          new_start = 0;
-        }
-        spl.insert( spl.end(), children.begin()+1, children.begin()+i+1 );
-        return NodeManager::currentNM()->mkNode( kind::PLUS, pl,
-                 NodeManager::currentNM()->mkNode( kind::STRING_STRIDOF, mkConcat( kind::STRING_CONCAT, spl ), node[1], NodeManager::currentNM()->mkConst( Rational(new_start) ) ) );
-      }else{
-        spl.insert( spl.end(), children.begin(), children.begin()+i+1 );
-        return NodeManager::currentNM()->mkNode( kind::STRING_STRIDOF, mkConcat( kind::STRING_CONCAT, spl ), node[1], node[2] );
+
+      // these rewrites are only possible if we will not return -1
```

Comment:
I don't quite understand this comment. This seems to be checking if node[1] is non-empty? What difference does this make if the str.contains above was rewritten to true?

Reply:
We don't return -1 for str.indexof( x, y, z ) if we can show that y is contained in str.substr( x, z, str.len( x ) ) and str.len( y ) > 0.  The latter is due to the semantics of indexof when the second argument is "".  

This is related to my reply to your comment above, this is the general case where we can strip a prefix:

z>str.len( x1 ) and str.len( y )>0 and str.contains( x2, y ) ----> true

implies 

str.index( str.++( x1, x2 ), y, w ) ---> str.len( x1 ) + str.indexof( x2, y, z-str.len(x1) )

Notice this requires all three conditions, which are check on lines 1922, 1917, 1894. The method on 1922 in fact generalizes this for dropping multiple components at once.